### ISSUE TITLE
Getting nested data when using Joins

### ISSUE THREAD

**Issue by thetutlage** [tgriesser/knex#882]
Currently if i use joins, it returns flat object, whereas returned object should be nested.

```
knex.select().from("users").leftJoin("user_addresses","users.id","user_addresses.user_id")
```

returned object is in following format

```javascript
{
 id: 3,
username: "somename",
email: "someemail",
street: "somestreet",
postcode: "somepostcode"
}
```

whereas i am expecting

```javascript
{
  id : 3,
  username : "somename"
  email: "someemail"
 user_addresses:[{
   street: "somestreet",
   postcode: "somepostcode"
}]
}
```

**Comment by rickytan** [comment]
All mysql query results is flat, you have to do it yourself!

```javascript
var _ = require('underscore');

knex.select().from("users").leftJoin("user_addresses","users.id","user_addresses.user_id").then(function(data) {
    return _.chain(data).groupBy(function(user) { reuturn user.id; }).map(function(users) {
      var user = _.chain(users).first().pick('id', 'username', 'email');
      var user_addresses = _.map(users, function(u) {
         return { 'street': u.street, 'postcode': u.postcode };
      });
      user.user_addresses = user_addresses;
      return user;
   }).value();
});
```

**Comment by drew-r** [comment]
Optionally nesting join results would be nice to have

**Comment by chrisbroome** [comment]
MySQL supports nested **tables** in query results, however this is not portable to other database drivers as far as I know. Use the `.options` method to specify driver-specific options per query.

``` JavaScript
knex.from('users')
  .leftJoin('user_addresses', 'users.id', '=', 'user_addresses.user_id')
  .options({nestTables: true})
  .then(results => {
    // do something with results ...
  });
```

`results` is an array of objects with each object having the following shape

``` JavaScript
{
  users: {
    id: 3,
    username: 'somename',
    email: 
  },
  user_addresses: {
    id: 1, // or whatever format your id is in
    user_id: 3,
    street: 'somestreet',
    postcode: 'somepostcode'
  }
}
```

Disclaimer: the following is a side note barely related to your question.

This is another good reason to name tables using singular form. The table name should represent the entity it is modeling, not the collection of entities that it stores. See this [StackOverflow](http://stackoverflow.com/a/5841297/436977) answer for more details.

**Comment by thetutlage** [comment]
Boom that's nice

**Comment by yohairosen** [comment]
For whom ever stumble across this in the future. Checkout [https://github.com/CoursePark/KnexNest](KnexNest(NestHydrationJS)). I've took the time to post it here since I don't really understand why this library isn't more popular - it's incredible.

**Comment by elhigu** [comment]
@yohairosen also objection.js and bookshelf allow fetching relations as nested data.

**Comment by yohairosen** [comment]
@elhigu These are ORMs. It's a complexity different approach than just mapping raw queries.

* full disclosure regarding objection.js would be nice :)

**Comment by elhigu** [comment]
@yohairosen yep fair enough. In those ORMish (not really traditional ORMs since they doesn't try to hide SQL queries behind additional layer of obscurity) things  you need to define relations too according to which nested data is generated, but on the other hand they work also through multiple level of queries so they can also fetch nested arrays of rows effectively like `village.houses.tenants` which are really inefficient to do with joins due to data amount explosion which joins causes (that is usually the case when I need nested data).

Agreed that for the simple case shown in that README and when that is all you need its 10 times simpler to just use knex nest.

What do you mean by full disclosure? Link to the relevant documentation http://vincit.github.io/objection.js/#eager  maybe?

**Comment by yohairosen** [comment]
I honestly did not know a single query with joins can be *less* efficient than round tripping to the DB. Unless it's being implemented behind the scenes with subqueries?

Do you have more information about this? I'm truly interested...

**Comment by elhigu** [comment]
@yohairosen when using joins with to-many relation, there is lots of data duplication in resulted rows and parsing that data comes really inefficient. Here is some analysis about it https://github.com/Vincit/objection.js/issues/73#issuecomment-224892996

**Comment by yohairosen** [comment]
@elhigu Thanks.

> Basically it is about the balance between data transfer + parsing in node and with roundtrip times.

So it seems. The discussion in the link doesn't conclude which method is better though. Seems like they have implemented a join query algorithm for you to choose (single query, no roundtrips). 

But, as the author mentions later in the discussion, it didn't seem to produce better performance. 

That being said, this wasn't a real world test since the DB runs on localhost which, at least in my own apps, is usually not the case. Running the DB on a remote machine would increase the roundtrip times greatly.

Furthermore, parsing the results can be done in workers, which I did not see the author mentions as a solution at all.

Some room for investigation, no doubt.

**Comment by elhigu** [comment]
There was lots of benchmarking/investigation involved I just didn't find the results anywhere with 2min search to the issues. AFAIK there was some numbers released also somewhere.

About using workers, parsing is done by DB drivers, they are quite optimized and some times native code, but afaik they shouldn't use multiple threads because it would not be good for overall performance of the system running hundreds of queries concurrently. Workers won't help overall performance only benefit would be that single query might be processed faster and overall benefit would be a bit worse (in theory) due to extra context switching. Base line is that parsing is driver's responsibility not knex / objection.

If there is 1-1 relation or 1-1-1-1-1-1 relations, join is usually better because they allow filtering whole query with joined attributes and because there are no -> many relations, there wont be overhead in data amount. One has to decide case-by-case which is better / more suitable.

btw. I have locally DB setup behind toxiproxy, which adds network lag simulation when running performance tests of that kind.

**Comment by vytautas-pranskunas-** [comment]
yohairosen KnexNest(NestHydrationJS) it isnice however it works only with underscores and is postgres returns data separated by dots it is useless :(

**Comment by mohaalak** [comment]
I write a module that does this even it generate types for it
https://github.com/mohaalak/nested-knex

**Comment by acro5piano** [comment]
I also wrote a module that does this (simple one using `whereIn`)

https://github.com/acro5piano/knex-with-relations

**Comment by herbola** [comment]
For postgres you can you 'json_agg' for one-to-many relations:


```js
knex
  .from('users')
  .leftJoin('user_addresses', 'users.id', '=', 'user_addresses.user_id')
  .select([
    'users.*',
    knext.raw(
      "case when count(user_addresses) = 0 then '[]' else json_agg(user_addresses.*) end",
    ),
  ])
  .groupBy(['users.id', 'user_addresses.user_id'])
  .first();
```

... which would output:

```json
{
  "id": 3,
  "username": "foo",
  "email": "foo",
  "user_addresses": [
    {
      "id": "?",
      "user_id": 3,
      "street": "foo",
      "postcode": "foo"
    }
  ]
}


```


if 'user_addresses' was a single entity (one-to-one), you could use following approach:

```js
knex
  .from('users')
  .leftJoin('user_addresses', 'users.id', '=', 'user_addresses.user_id')
  .select(['users.*', knex.raw('(json_agg(user_addresses.*) ->> 0)::json')])
  .groupBy(['users.id', 'user_addresses.user_id'])
  .first();
```

... which would output:
```json
{
  "id": 3,
  "username": "foo",
  "email": "foo",
  "user_addresses": {
    "id": "?",
    "user_id": 3,
    "street": "foo",
    "postcode": "foo"
  }
}

```

**Comment by mercmobily** [comment]
Hi there!

Knex is a SQL query builder, not an ORM. It deliberately does not transform flat SQL result sets into nested objects. This is a design decision that will not change.

As stated by maintainer @tgriesser in https://github.com/knex/knex/issues/61#issuecomment-61946622: "this is a limitation of sql... and not something knex will try to handle for you provide out of the box".

For detailed solutions (using ORMs like Objection.js, database-specific JSON functions, manual transformation, or separate queries), please see the master documentation issue: #6289

This issue will be closed as a duplicate of #6289.
